capstone/capapi/permissions.py

```diff
@@ -18,9 +19,41 @@ def has_permission(self, request, view):
             # boot if user tries anything but `GET`, `OPTIONS`, or `HEAD`
             return False
 
-        # give away metadata to everyone!
-        if not request.query_params.get('type') == 'download':
+        # check whether user wants to get full case text
+        context = view.get_renderer_context()
+        full_case = context.get('request').query_params.get('full_case', 'false').lower()
+        if full_case != 'true':
+            # serve metadata to everyone!
             return True
+        kwargs = context.get('kwargs')
+        # get blacklisted case count
+        cite = context.get('request').query_params.get('cite', None)
+
+        if view.lookup_field in kwargs and '-' in slugify(kwargs[view.lookup_field]):
+            # assume this is a lookup using a citation if hyphen is present
+            cases = view.queryset.filter(citation__normalized_cite=slugify(kwargs[view.lookup_field]),
+                                         jurisdiction__whitelisted=False)
+        elif cite:
+            cases = view.queryset.filter(citation__normalized_cite=slugify(cite),
+                                         jurisdiction__whitelisted=False)
         else:
-            # make sure user has auth token before proceeding if they want to download
-            return request.auth
+            cases = view.queryset.filter(**kwargs, jurisdiction__whitelisted=False)
+
+        blacklisted_count = cases.count()
+
+        # If the request contains only whitelisted jurisdictions
+        # serve up whether the user is signed in or not
+        if not blacklisted_count:
+            return True
+
+        # from here forth one cannot be anonymous
+        if request.user.is_anonymous:
+            return False
+
+        is_allowed = request.user.case_allowance_remaining >= blacklisted_count
+
```

Comment:
This is logic that used to exist in the view's CaseViewSet download function. It has been moved here because I think permissions logic appropriately should live in the permissions module. I am still conflicted about this, though, because this means that we're doubling the lookups, as DRF's permissions only (I think) allows the returning of a Boolean. This means that if a user *does* have permissions, this look up will happen again in the CaseViewSet. I'm hoping django's and psql's caching might be helpful to speed this up, but I don't know this for sure. What do you think?

Reply:
I think maybe we were going to handle this in a different way that avoids this problem. Rather than require login if the request includes blacklisted cases, and refusing to return if they ask for too many, we were going to filter the results to return only allowed items. So like logged-out users get

```
{
    "casebody": null,
    "casebody_missing_reason": "Please log in to retrieve casebody field for Massachusetts."
}
```

And logged in users who have exceeded their daily limit would get a different casebody_missing_reason.

Some issues to kick around with this approach too ...